### PR TITLE
Throw InvalidDataException on duplicate dictionary key

### LINKED ISSUE #4455: [Audit-Medium] duplicate request-context keys are misclassified as in…
> **AI-generated audit finding** — this issue was opened from an automated security/correctness audit. It **has not been triaged by a human yet**; verify the reasoning, reproducibility, and severity before acting on it.

### Medium: duplicate request-context keys are misclassified as internal server failures — **CONFIRMED**

Affected code:

- [src/IceRpc.RequestContext/RequestContextMiddleware.cs:32-43](https://github.com/icerpc/icerpc-csharp/blob/150d6ee01de9609eb037963cf2897df190c23809/src/IceRpc.RequestContext/RequestContextMiddleware.cs#L32) — uses `new Dictionary<string, string>(size)` as the factory for both protocols
- [src/IceRpc/Ice/Codec/IceDecoderExtensions.cs:51](https://github.com/icerpc/icerpc-csharp/blob/150d6ee01de9609eb037963cf2897df190c23809/src/IceRpc/Ice/Codec/IceDecoderExtensions.cs#L51), [src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs:52](https://github.com/icerpc/icerpc-csharp/blob/150d6ee01de9609eb037963cf2897df190c23809/src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs#L52) — `dictionary.Add(new KeyValuePair<TKey, TValue>(key, value));`
- [src/IceRpc/Internal/IceRpcProtocolConnection.cs:1210](https://github.com/icerpc/icerpc-csharp/blob/150d6ee01de9609eb037963cf2897df190c23809/src/IceRpc/Internal/IceRpcProtocolConnection.cs#L1210), [src/IceRpc/Internal/IceProtocolConnection.cs:1051](https://github.com/icerpc/icerpc-csharp/blob/150d6ee01de9609eb037963cf2897df190c23809/src/IceRpc/Internal/IceProtocolConnection.cs#L1051) — protocol layers catch `InvalidDataException` as malformed-input; other exceptions map to `InternalError`

Verification:

Confirmed. `ICollection<KeyValuePair<TKey, TValue>>.Add` on a `Dictionary<string, string>` routes to `IDictionary.Add(key, value)`, which throws `ArgumentException` on a duplicate key (C# spec / `Dictionary<TKey,TValue>.Add` docs). `ArgumentException` is not `InvalidDataException`, so the protocol layer's malformed-input branch doesn't catch it and the dispatch fails with `StatusCode.InternalError` instead of a clean malformed-input error.

Impact:

- A trivially malformed peer payload (duplicate key in the context dictionary) is reported as a server bug.
- Error classification leaks implementation details and complicates incident triage.

Recommendation:

- Wrap the `DecodeDictionary` calls in a `try/catch (ArgumentException ex) { throw new InvalidDataException("duplicate key in request context", ex); }`.
- Alternatively, decode into a `KeyValuePair<string, string>[]` first, detect duplicates explicitly, and translate to `InvalidDataException` with a useful message that identifies the duplicate key.
- Add tests for duplicate keys on both `ice` and `icerpc`.

Status: **Valid, Medium severity**.

---

**Source report:** [src-IceRpc.RequestContext-audit-2026-04-14.md](https://github.com/icerpc/icerpc-csharp-audit/blob/main/src-IceRpc.RequestContext-audit-2026-04-14.md) (finding `duplicate request-context keys are misclassified as internal server failures — **CONFIRMED**`)

**Severity (auditor-assigned):** Medium

### PR DESCRIPTION
**Pull request by pepone**
Fixes #4455.

### CODE REVIEW

File: src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs [hunk 1]
```diff
@@ -49,7 +49,16 @@ public static TDictionary DecodeDictionary<TDictionary, TKey, TValue>(
...
+                try
+                {
+                    dictionary.Add(new KeyValuePair<TKey, TValue>(key, value));
+                }
+                catch (ArgumentException exception)
+                {
+                    throw new InvalidDataException(
+                        $"Received dictionary with duplicate key '{key}'.",
+                        exception);
+                }
```
[the remark this answers is not in the document]
Comment (pepone) [reply src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs:59]: null keys are not allow with Slice encoding

File: src/IceRpc/Ice/Codec/IceDecoderExtensions.cs [hunk 2]
```diff
@@ -48,7 +48,16 @@ public static TDictionary DecodeDictionary<TDictionary, TKey, TValue>(
...
-                dictionary.Add(new KeyValuePair<TKey, TValue>(key, value));
+                try
+                {
+                    dictionary.Add(new KeyValuePair<TKey, TValue>(key, value));
+                }
+                catch (ArgumentException exception)
+                {
+                    throw new InvalidDataException(
+                        $"Received dictionary with duplicate key '{key}'.",
+                        exception);
```
[the remark this answers is not in the document]
Comment (pepone) [reply src/IceRpc/Ice/Codec/IceDecoderExtensions.cs]: I find `ICollection<KeyValuePair<` a bit problematic, `IDictionary` is a better fit for Slice dictionaries because it has the same no duplicate semantics.

**Review by externl — APPROVED** [review]

File: src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs [hunk 3]
```diff
@@ -28,6 +28,11 @@ public static void CheckEndOfBuffer(this ref SliceDecoder decoder)
     /// <param name="keyDecodeFunc">The decode function for each key of the dictionary.</param>
     /// <param name="valueDecodeFunc">The decode function for each value of the dictionary.</param>
     /// <returns>The dictionary decoded by this decoder.</returns>
+    /// <remarks>Duplicate-key detection depends on the collection returned by <paramref name="dictionaryFactory" />.
```
Comment (bernardnormier) [inline src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs]: I would do simpler and specify:
 - TKey can't be null
 - TDictionary must be an `IDictionary<TKey, TValue>`

This way, the semantics of ArgumentException is well defined.

File: src/IceRpc/Ice/Codec/IceDecoderExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply src/IceRpc/Ice/Codec/IceDecoderExtensions.cs]: As per my other comment, we could switch to `IDictionary` for parameters (already done for fields). We'll need to update the doc. 

I think for the Ice IDL/encoding, the 3 options (Dictionary, SortedList, SortedDictionary) are all IDictionary already.

File: src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs (hunk 3 above)
Comment (pepone) [reply src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs]: I'm removing this remarks, I will address this in a separate PR see https://github.com/icerpc/icerpc-csharp/issues/4519

Just keep the new test and exception handling for this PR.

**Review by bernardnormier — APPROVED** [review]

File: src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs (hunk 3 above)
Comment (InsertCreativityHere) [reply src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs]: Should we just do #4519 instead then?

Comment (pepone) [reply src/ZeroC.Slice.Codec/SliceDecoderExtensions.cs]: I merge this because the test and the exception handling is still the same.
